proposals/0002-gep.md

```diff
@@ -0,0 +1,140 @@
+---
+MEP: 0002
+Title: Unifed API
+Discussion: Link
+Implementation: Link
+---
+
+# TITLE: Unified API
+
+## Abstract
+
+This GEP discusses the unified API that handles the interaction with LLM providers. This represents the feature set for MVP.
+
+## Motivation
+
+
+The Unified API abstracts much of the overhead of interacting with LLM providers.
+It stores all the provider specific logic in a single place and allowing developers to focus on the core logic. It also introduces durability to LLM apps. Additionally, unifying the API allows optimizations to be leveraged such as latency and cost improvements.
+
+
+### Requirements
+
+- R1: User should be able to specify specific models or select auto mode (selection handled by Glide)
+- R2: Should be a Hertz Client
+- R3: Should take a user request as input
+- R4: Handles all provider specific logic
+- R5: Easily maintained
+- R6: API schemas must unify common LLM request params (e.g. topP, N, etc)
+- R7: API routes must unify common LLM endpoints/API
+- R8: API schemas should allow to customize provider-specific params (e.g. a different messages for OpenAI/Anthropic)
+
+
+## Design
+
+
+- A Hertz client that handles an incoming request and returns a reponse from that provider.
+- Provider specific clients
+
+### Routes
+```yaml
+routes:
+    chat: /chat/{pool-id}
+    complete: /complete/{pool-id} #this is also know as generate in some providers
+    multi-modal: /multimodal/{pool-id} # future route, not for MVP
+    embedding: /embedding/{pool-id} # future route, not for MVP
+```
+
+#### User Request Schema for Chat Route
+
+Below is an example user request to the Chat route. The other routes will look very similar, if not identical. The provider specific params will look different.
+
+Example Request Body:
+
+
+```json
+{
+  "message": "tell me a joke",
+}
+```
+
+### Provider Request Schemas
+
+Request schema from OpenAI Client to OpenAI API
+
+```json
+
+{
+    "model": "gpt-3.5-turbo",
+    "messages": [
+      {
+        "role": "system",
+        "content": "You are a helpful assistant."
+      },
+      {
+        "role": "user",
+        "content": "tell me a joke"
+      }
+    ],
+    "temperature": 0.7,
+    "max_tokens": 100,
+}
+
+```
+
+### Response Schema For Chat
+
+For MVP, lets just return the response from the provider. Once we get feedback we can incorporate more metedata from the gateway about the request.
```

Comment:
In my mind, the response schemas are a part of the Unified API. It doesn't have to be something complex, but we could at least separate the main part (e.g. the next message the model replied based on the chat history) from metadata which is essentially pretty much everything else in a free map like form (or `map[string]interface` in go).

I'm sure though there has to be more common things we could unify.

Reply:
Another thought: we probably cannot just leave it as it really, because what would happen to client if we have the following case:
- An application uses a pool with OpenAI and Anthropic as a fallback
- The app does request to Glide Unified API and returns a response in OpenAI format (it happens to be available at the point)
- Then, at some point, OpenAI API goes down and the app suddenly starts to receive Anthropic response schema

Q: Should this difference handling be on the application side? I feel like not even for MVP.

Reply:
Yep, yep this is a good point. I'll update it with something more durable